core/src/main/java/com/linecorp/armeria/server/Server.java

```diff
@@ -655,52 +657,79 @@ private void doStop(CompletableFuture<Void> future,
-                                }
-
-                                // Boss groups have been terminated completely.
+                        workerShutdownFuture.addListener(unused7 -> {
+                            final Set<EventLoopGroup> bossGroups =
+                                    Server.this.serverChannels.stream()
+                                                              .map(ch -> ch.eventLoop().parent())
+                                                              .collect(toImmutableSet());
+
+                            // If started to shutdown before initializing a boss group,
+                            // complete the future immediately.
+                            if (bossGroups.isEmpty()) {
                                 finishDoStop(future);
+                                return;
+                            }
+
+                            // Shut down all boss groups and wait until they are terminated.
+                            final AtomicInteger remainingBossGroups = new AtomicInteger(bossGroups.size());
+                            bossGroups.forEach(bossGroup -> {
+                                bossGroup.shutdownGracefully();
+                                bossGroup.terminationFuture().addListener(unused8 -> {
+                                    if (remainingBossGroups.decrementAndGet() != 0) {
+                                        // There are more boss groups to terminate.
+                                        return;
+                                    }
+
+                                    // Boss groups have been terminated completely.
+                                    finishDoStop(future);
+                                });
                             });
                         });
-                    });
 
+                        return null;
+                    });
                     return null;
                 });
 
                 return null;
             });
         }
 
+        private CompletableFuture<Void> shutdownServerHandlers() {
+            if (config.gracefulShutdown().timeout().isZero()) {
+                return UnmodifiableFuture.completedFuture(null);
+            }
+
+            final Set<Channel> children = connectionLimitingHandler.children();
+            final List<CompletableFuture<Void>> closeFutures = new ArrayList<>(children.size());
+            for (Channel ch : children) {
+                final HttpServerHandler serverHandler = ch.pipeline().get(HttpServerHandler.class);
+                if (serverHandler != null) {
+                    closeFutures.add(serverHandler.shutdown(ch));
+                }
+            }
+
+            // The future returned by shutdown() will be always completed successfully.
+            final CompletableFuture<List<Void>> combined = CompletableFutures.allAsList(closeFutures);
+            config.workerGroup().schedule(() -> {
+                combined.complete(ImmutableList.of());
```

Comment:
This won't wait all `closeFutures` to be completed if completing `closeFutures` takes more than 1 second.

Reply:
It was intentional. A graceful timeout has passed since the stop process started. 1 second is the last change/time to send shutting-down responses before the connections are closed.

Reply:
If it was intentional, let's not call `CompletableFutures.allAsList(closeFutures);` because it's useless. We can do:
```java
final CompletableFuture<Void> future = new CompletableFuture<>();
config.workerGroup().schedule(() -> {
    future.complete();
}, 1, TimeUnit.SECONDS);
return future;
```

Reply:
Oops, NVM. It is still needed because the future can be completed earlier.